Source/Core/VideoBackends/D3D12/ShaderCache.cpp

```diff
@@ -0,0 +1,388 @@
+	s_gs_disk_cache.OpenAndRead(gs_cache_filename, gs_inserter);
+
+	ShaderCacheInserter<SHADER_STAGE_PIXEL_SHADER> ps_inserter;
+	s_ps_disk_cache.OpenAndRead(ps_cache_filename, ps_inserter);
+
+	ShaderCacheInserter<SHADER_STAGE_VERTEX_SHADER> vs_inserter;
+	s_vs_disk_cache.OpenAndRead(vs_cache_filename, vs_inserter);
+
+	// Clear out disk cache when debugging shaders to ensure stale ones don't stick around..
+	if (g_Config.bEnableShaderDebugging)
+		Clear();
+}
+
+void ShaderCache::Clear()
+{
+
+}
+
+void ShaderCache::Shutdown()
+{
+	for (auto& iter : s_gs_bytecode_cache)
+		SAFE_DELETE(iter.second.pShaderBytecode);
+
+	for (auto& iter : s_ps_bytecode_cache)
+		SAFE_DELETE(iter.second.pShaderBytecode);
+
+	for (auto& iter : s_vs_bytecode_cache)
+		SAFE_DELETE(iter.second.pShaderBytecode);
+
+	s_gs_bytecode_cache.clear();
+	s_ps_bytecode_cache.clear();
+	s_vs_bytecode_cache.clear();
+
+	s_gs_disk_cache.Sync();
+	s_gs_disk_cache.Close();
+	s_ps_disk_cache.Sync();
+	s_ps_disk_cache.Close();
+	s_vs_disk_cache.Sync();
+	s_vs_disk_cache.Close();
+
+	if (g_Config.bEnableShaderDebugging)
+	{
+		s_gs_hlsl_cache.clear();
+		s_ps_hlsl_cache.clear();
+		s_vs_hlsl_cache.clear();
+	}
+
+	s_geometry_uid_checker.Invalidate();
+	s_pixel_uid_checker.Invalidate();
+	s_vertex_uid_checker.Invalidate();
+}
+
+void ShaderCache::LoadAndSetActiveShaders(DSTALPHA_MODE ps_dst_alpha_mode, u32 gs_primitive_type)
+{
+	switch (gs_primitive_type)
+	{
+	case PRIMITIVE_TRIANGLES:
+		s_current_primitive_topology = D3D12_PRIMITIVE_TOPOLOGY_TYPE_TRIANGLE;
+		break;
+	case PRIMITIVE_LINES:
+		s_current_primitive_topology = D3D12_PRIMITIVE_TOPOLOGY_TYPE_LINE;
+		break;
+	case PRIMITIVE_POINTS:
+		s_current_primitive_topology = D3D12_PRIMITIVE_TOPOLOGY_TYPE_POINT;
+		break;
+	default:
+		CHECK(0, "Invalid primitive type.");
+		break;
+	}
+
+	GeometryShaderUid gs_uid = GetGeometryShaderUid(gs_primitive_type, API_D3D);
+	PixelShaderUid ps_uid = GetPixelShaderUid(ps_dst_alpha_mode, API_D3D);
+	VertexShaderUid vs_uid = GetVertexShaderUid(API_D3D);
+
+	bool gs_changed = gs_uid != s_last_geometry_shader_uid;
+	bool ps_changed = ps_uid != s_last_pixel_shader_uid;
+	bool vs_changed = vs_uid != s_last_vertex_shader_uid;
+
+	if (!gs_changed && !ps_changed && !vs_changed)
+	{
+		return;
+	}
+
+	// A Uid has changed, so the PSO will need to be reset at next ApplyState.
+	D3D::command_list_mgr->m_dirty_pso = true;
+
+	if (g_ActiveConfig.bEnableShaderDebugging)
+	{
+		ShaderCode code;
+		if (gs_changed)
+		{
+			code = GenerateGeometryShaderCode(gs_primitive_type, API_D3D);
+			s_geometry_uid_checker.AddToIndexAndCheck(code, gs_uid, "Geometry", "g");
+		}
+
+		if (ps_changed)
+		{
+			code = GeneratePixelShaderCode(ps_dst_alpha_mode, API_D3D);
+			s_pixel_uid_checker.AddToIndexAndCheck(code, ps_uid, "Pixel", "p");
+		}
+		
+		if (vs_changed)
+		{
+			code = GenerateVertexShaderCode(API_D3D);
+			s_vertex_uid_checker.AddToIndexAndCheck(code, vs_uid, "Vertex", "v");
+		}
+	}
+
+	if (gs_changed)
```

Comment:
Each of these if statement blocks (gs_changed, vs_changed, ps_changed) should be their own functions.